File: DiscriminatedUnions/LegacyExamples.ts
// Alternative TypeScript style using manual type guards and if/else branching.
// Це приклад до сучасного патерн-матчінгу, коли в коді вручну перевіряють `kind`.

type Circle = { kind: "circle"; radius: number };
type Rectangle = { kind: "rectangle"; width: number; height: number };
type Triangle = { kind: "triangle"; baseLength: number; height: number };

type Shape = Circle | Rectangle | Triangle;

function isCircle(shape: Shape): shape is Circle {
  return shape.kind === "circle";
}

function isRectangle(shape: Shape): shape is Rectangle {
  return shape.kind === "rectangle";
}

function isTriangle(shape: Shape): shape is Triangle {
  return shape.kind === "triangle";
}

function area(shape: Shape): number {
  if (isCircle(shape)) {
    return Math.PI * shape.radius * shape.radius;
  }

  if (isRectangle(shape)) {
    return shape.width * shape.height;
  }

  if (isTriangle(shape)) {
    return (shape.baseLength * shape.height) / 2;
  }

  throw new Error("Unknown shape");
}

function describe(shape: Shape): string {
  if (isCircle(shape)) {
    return `Circle with radius ${shape.radius}`;
  }

  if (isRectangle(shape)) {
    return `Rectangle ${shape.width}x${shape.height}`;
  }

  if (isTriangle(shape)) {
    return `Triangle base ${shape.baseLength} height ${shape.height}`;
  }

  throw new Error("Unknown shape");
}

const shapes: Shape[] = [
  { kind: "circle", radius: 3 },
  { kind: "rectangle", width: 4, height: 5 },
  { kind: "triangle", baseLength: 6, height: 4 },
];

for (const shape of shapes) {
  console.log(`${describe(shape)} -> area = ${area(shape).toFixed(2)}`);
}

File: DiscriminatedUnions/LegacyExamples.cs
// Alternative C# style before records and richer pattern matching.
// Тут ми використовуємо класичну об'єктно-орієнтовану ієрархію типів.

using System;
using System.Collections.Generic;

namespace DiscriminatedUnions.Legacy
{
    public abstract class Shape
    {
        public abstract double Area();
        public abstract string Describe();
    }

    public class Circle : Shape
    {
        public double Radius { get; }

        public Circle(double radius) => Radius = radius;

        public override double Area() => Math.PI * Radius * Radius;
        public override string Describe() => $"Circle with radius {Radius}";
    }

    public class Rectangle : Shape
    {
        public double Width { get; }
        public double Height { get; }

        public Rectangle(double width, double height)
        {
            Width = width;
            Height = height;
        }

        public override double Area() => Width * Height;
        public override string Describe() => $"Rectangle {Width}x{Height}";
    }

    public class Triangle : Shape
    {
        public double BaseLength { get; }
        public double Height { get; }

        public Triangle(double baseLength, double height)
        {
            BaseLength = baseLength;
            Height = height;
        }

        public override double Area() => BaseLength * Height / 2.0;
        public override string Describe() => $"Triangle base {BaseLength} height {Height}";
    }

    class Program
    {
        static void Main()
        {
            var shapes = new List<Shape>
            {
                new Circle(3.0),
                new Rectangle(4.0, 5.0),
                new Triangle(6.0, 4.0)
            };

            foreach (var shape in shapes)
            {
                Console.WriteLine($"{shape.Describe()} -> area = {shape.Area():F2}");
            }
        }
    }
}
